File: HealthcareData/scripts/cleanDoctors.js
/**
 * cleanDoctors.js
 * Scaffolding script to clean raw doctors datasets.
 * Usage: node scripts/cleanDoctors.js <raw_file_name>
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const WORKSPACE_DIR = path.join(__dirname, '..');
const RAW_DIR = path.join(WORKSPACE_DIR, 'raw', 'doctors');
const CLEANED_DIR = path.join(WORKSPACE_DIR, 'cleaned', 'doctors');
const LOG_FILE = path.join(WORKSPACE_DIR, 'logs', 'doctor_clean.log');

function log(msg) {
  const timestamp = new Date().toISOString();
  const formatted = `[${timestamp}] ${msg}\n`;
  console.log(msg);
  fs.appendFileSync(LOG_FILE, formatted);
}

async function run() {
  const args = process.argv.slice(2);
  const inputFileName = args[0];

  if (!inputFileName) {
    log('ERROR: Please specify a raw file name inside raw/doctors/. Example: node scripts/cleanDoctors.js doctors_list.json');
    process.exit(1);
  }

  const inputPath = path.join(RAW_DIR, inputFileName);
  if (!fs.existsSync(inputPath)) {
    log(`ERROR: Raw file not found at: ${inputPath}`);
    process.exit(1);
  }

  log(`Starting cleanup process for: ${inputFileName}`);

  try {
    const rawData = fs.readFileSync(inputPath, 'utf-8');
    let records = [];

    if (inputFileName.endsWith('.json')) {
      records = JSON.parse(rawData);
    } else {
      const lines = rawData.split('\n').filter(Boolean);
      const headers = lines[0].split(',').map(h => h.trim());
      records = lines.slice(1).map(line => {
        const values = line.split(',').map(v => v.trim());
        const obj = {};
        headers.forEach((h, i) => {
          obj[h] = values[i] || '';
        });
        return obj;
      });
    }

    log(`Loaded ${records.length} records from raw dataset.`);

    const cleanedRecords = [];
    const seenEmails = new Set();
    let duplicates = 0;
    let emptyEmails = 0;

    for (const record of records) {
      const email = (record.email || record.Email || '').trim().toLowerCase();
      const firstName = (record.firstName || record.First_Name || record.Name || '').trim();

      if (!email) {
        emptyEmails++;
        continue;
      }

      if (seenEmails.has(email)) {
        duplicates++;
        continue;
      }

      seenEmails.add(email);

      const cleaned = {
        firstName,
        lastName: (record.lastName || record.Last_Name || '').trim(),
        email,
        specialization: (record.specialization || record.Specialty || 'General Practitioner').trim(),
        experienceYears: parseInt(record.experienceYears || record.Experience || 0, 10),
        contactNumber: (record.phone || record.Contact_No || '').trim().replace(/[^\d+]/g, ''),
        consultationFee: parseInt(record.consultationFee || record.Fee || 500, 10),
      };

      cleanedRecords.push(cleaned);
    }

    log(`Cleanup summary:`);
    log(`- Total original records: ${records.length}`);
    log(`- Cleaned records output: ${cleanedRecords.length}`);
    log(`- Removed duplicates: ${duplicates}`);
    log(`- Removed empty emails: ${emptyEmails}`);

    const outputFileName = `cleaned_${inputFileName.replace(/\.[^/.]+$/, "")}.json`;
    const outputPath = path.join(CLEANED_DIR, outputFileName);
    fs.writeFileSync(outputPath, JSON.stringify(cleanedRecords, null, 2), 'utf-8');

    log(`SUCCESS: Cleaned records successfully written to: ${outputPath}`);
  } catch (error) {
    log(`CRITICAL ERROR during doctor cleanup: ${error.message}`);
    process.exit(1);
  }
}

run();

File: HealthcareData/scripts/cleaning/clean_doctors.py
#!/usr/bin/env python3
import os
import pandas as pd
from HealthcareData.config import RAW_DIR, CLEANED_DIR
from HealthcareData.scripts.logger_helper import PipelineLogger
from HealthcareData.scripts.validation.validate_dataset import validate

def clean_doctors():
    plog = PipelineLogger("Clean_Doctors", "clean_doctors.log")
    plog.info("Starting clean doctors dataset pipeline...")
    
    input_dir = os.path.join(RAW_DIR, "doctors")
    output_dir = os.path.join(CLEANED_DIR, "doctors")
    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, "doctors.csv")
    
    # Check for files
    files = [f for f in os.listdir(input_dir) if not f.startswith('.')] if os.path.exists(input_dir) else []
    
    if len(files) == 0:
        plog.warning("Dataset structure warning: raw/doctors/ directory is empty. Generating empty template doctors.csv.")
        
        # Create empty template matching required columns in config.py
        required_cols = ["firstName", "lastName", "email", "specialization", "experienceYears", "contactNumber", "consultationFee"]
        df = pd.DataFrame(columns=required_cols)
        
        # Add one mock doctor to make it look clean if needed, or leave it empty. Let's add an empty df.
        df.to_csv(output_path, index=False)
        plog.info(f"Successfully saved empty template doctors.csv to: {output_path}")
        plog.finalize()
        return

    # If files exist, clean the first one
    input_file = files[0]
    input_path = os.path.join(input_dir, input_file)
    plog.info(f"Found raw doctors file: {input_file}. Starting cleanup...")

    try:
        if input_file.endswith('.json'):
            df = pd.read_json(input_path)
        else:
            df = pd.read_csv(input_path)
            
        plog.records_processed = len(df)
        plog.info(f"Loaded {len(df)} records.")

        # Clean column names
        df.columns = [col.strip() for col in df.columns]

        # Rename common aliases
        column_mapping = {
            "First_Name": "firstName",
            "Name": "firstName",
            "Last_Name": "lastName",
            "Email": "email",
            "Specialty": "specialization",
            "Experience": "experienceYears",
            "phone": "contactNumber",
            "Contact_No": "contactNumber",
            "Fee": "consultationFee"
        }
        df = df.rename(columns=column_mapping)

        # Remove duplicate rows
        duplicates = df.duplicated().sum()
        if duplicates > 0:
            df = df.drop_duplicates()
            plog.duplicates_removed = duplicates
            plog.info(f"Removed {duplicates} duplicate rows.")

        # Fill missing values
        required_cols = ["firstName", "lastName", "email", "specialization", "experienceYears", "contactNumber", "consultationFee"]
        for col in required_cols:
            if col not in df.columns:
                df[col] = None
        df = df[required_cols]

        df["firstName"] = df["firstName"].fillna("Unknown")
        df["lastName"] = df["lastName"].fillna("Doctor")
        df["email"] = df["email"].fillna("info@hospital.com")
        df["specialization"] = df["specialization"].fillna("General Practitioner")
        df["experienceYears"] = pd.to_numeric(df["experienceYears"], errors='coerce').fillna(1).astype(int)
        df["contactNumber"] = df["contactNumber"].astype(str).str.replace(r'[^\d+]', '', regex=True).fillna("9876543210")
        df["consultationFee"] = pd.to_numeric(df["consultationFee"], errors='coerce').fillna(500).astype(int)

        # Validate
        validate(df, "doctors", plog)

        # Save
        df.to_csv(output_path, index=False)
        plog.info(f"Successfully saved cleaned file to: {output_path}")

    except Exception as e:
        plog.error(f"Error during doctors cleaning: {str(e)}")
        
    plog.finalize()

if __name__ == "__main__":
    clean_doctors()
